src/java/org/apache/cassandra/utils/bytecomparable/ByteSource.java

```diff
@@ -0,0 +1,813 @@
+    // Marker used to present null values represented by empty buffers (e.g. by Int32Type)
+    int NEXT_COMPONENT_EMPTY = 0x3F;
+    int NEXT_COMPONENT_EMPTY_REVERSED = 0x41;
+    // Marker for null components in tuples, maps, sets and clustering keys.
+    int NEXT_COMPONENT_NULL = 0x3E;
+    // Default terminator byte in sequences. Smaller than NEXT_COMPONENT_NULL, but larger than LT_NEXT_COMPONENT to
+    // ensure lexicographic compares go in the correct direction
+    int TERMINATOR = 0x38;
+    // These are special endings, for exclusive/inclusive bounds (i.e. smaller than anything with more components, bigger than anything with more components)
+    int LT_NEXT_COMPONENT = 0x20;
+    int GT_NEXT_COMPONENT = 0x60;
+
+    // Special value for components that should be excluded from the normal min/max span. (static rows)
+    int EXCLUDED = 0x18;
+
+    /**
+     * Reinterprets a byte buffer as a byte-comparable source that has 0s escaped and finishes in an escape.
+     * This provides a weakly-prefix-free byte-comparable version of the content to use in sequences.
+     * (See ByteSource.BufferReinterpreter/Multi for explanation.)
+     */
+    static <V> ByteSource of(ValueAccessor<V> accessor, V data, Version version)
+    {
+        return new AccessorReinterpreter<>(accessor, data, version);
+    }
+
+    /**
+     * Reinterprets a byte buffer as a byte-comparable source that has 0s escaped and finishes in an escape.
+     * This provides a weakly-prefix-free byte-comparable version of the content to use in sequences.
+     * (See ByteSource.BufferReinterpreter/Multi for explanation.)
+     */
+    static ByteSource of(ByteBuffer buf, Version version)
+    {
+        return new BufferReinterpreter(buf, version);
+    }
+
+    /**
+     * Reinterprets a byte array as a byte-comparable source that has 0s escaped and finishes in an escape.
+     * This provides a prefix-free byte-comparable version of the content to use in sequences.
+     * (See ByteSource.BufferReinterpreter/Multi for explanation.)
+     */
+    static ByteSource of(byte[] buf, Version version)
+    {
+        return new ReinterpreterArray(buf, version);
+    }
+
+    /**
+     * Reinterprets a memory range as a byte-comparable source that has 0s escaped and finishes in an escape.
+     * This provides a weakly-prefix-free byte-comparable version of the content to use in sequences.
+     * (See ByteSource.BufferReinterpreter/Multi for explanation.)
+     */
+    static ByteSource of(long address, int length, ByteComparable.Version version)
+    {
+        return new MemoryReinterpreter(address, length, version);
+    }
+
+    /**
+     * Combines a chain of sources, turning their weak-prefix-free byte-comparable representation into the combination's
+     * prefix-free byte-comparable representation, with the included terminator character.
+     * For correctness, the terminator must be within MIN-MAX_SEPARATOR and different from NEXT_COMPONENT+/-1.
```

Comment:
nit: Would it make sense to codify this as an assertion, or is that problematic because of how hot it is combined with our inability to disable assertions? :)